### PR TITLE
add subgraphpaths restful api

### LINKED ISSUE #39: hugegraph在进行出度遍历时存在性能问题
描述：从一个已知节点出发，向外（out）遍历，直至不存在叶子节点为止
语句：g.V().hasId(\"xxx\").repeat(out(\"look\")).until(outE(\"look\").count().is(0)).path()
数据：3000多条边+3000多个顶点；  **上述语句执行花了12s**+  请问什么原因 可否优化

### PR DESCRIPTION
**Pull request by zhoney**
fix #39

Change-Id: I41f82b73cee0ce657b3bcaaf1a68cf3479788fbf

### CODE REVIEW

File: hugegraph-api/src/main/java/com/baidu/hugegraph/serializer/JsonSerializer.java [hunk 1]
```diff
@@ -194,6 +198,33 @@ public String writePaths(String name, Collection<HugeTraverser.Path> paths,
         return writeList(name, pathList);
     }
 
+    @Override
+    public String writeSubGraphPaths(
+                  MultivaluedMap<Boolean, HugeTraverser.Path> paths,
+                  boolean withCrossPoint) {
+        try (ByteArrayOutputStream out = new ByteArrayOutputStream(LBUF_SIZE)) {
+            out.write("{\"loop path\": ".getBytes(API.CHARSET));
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/serializer/JsonSerializer.java]: loop_path

File: hugegraph-api/src/main/java/com/baidu/hugegraph/serializer/JsonSerializer.java [hunk 2]
```diff
@@ -194,6 +198,33 @@ public String writePaths(String name, Collection<HugeTraverser.Path> paths,
...
+            this.writePathsList(out, paths.get(true));
+            out.write(",\"leaf path\": ".getBytes(API.CHARSET));
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/serializer/JsonSerializer.java]: leaf_path

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/SubGraphPaths.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        Id source = VertexAPI.checkAndParseVertexId(sourceV);
+        Directions dir = Directions.convert(EdgeAPI.parseDirection(direction));
+
+        HugeGraph g = graph(manager, graph);
+
+        HugeTraverser traverser = new HugeTraverser(g);
+        MultivaluedMap paths = traverser.subGraphPaths(source, dir, edgeLabel,
+                                                       depth, degree, capacity,
+                                                       limit);
+        return manager.serializer(g).writeSubGraphPaths(paths, false);
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/SubGraphPaths.java]: prefer return a list of path

File: hugegraph-api/src/main/java/com/baidu/hugegraph/serializer/JsonSerializer.java [hunk 4]
```diff
@@ -194,6 +198,33 @@ public String writePaths(String name, Collection<HugeTraverser.Path> paths,
...
+        }
+    }
+
+    private void writePathsList(ByteArrayOutputStream out,
+                                List<HugeTraverser.Path> paths)
+                                throws IOException {
+        if (paths == null) {
+            out.write("[]".getBytes(API.CHARSET));
+        } else {
+            this.writer.writeObject(out, paths.stream().map(p -> p.toMap(false))
```
[the remark this answers is not in the document]
Comment (zhoney) [reply hugegraph-api/src/main/java/com/baidu/hugegraph/serializer/JsonSerializer.java]: done

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeTraverser.java [hunk 5]
```diff
@@ -555,6 +593,94 @@ public boolean reachLimit() {
...
+            this.degree = degree;
+            this.capacity = capacity;
+            this.limit = limit;
+            this.pathCount = 0L;
+        }
+
+        /**
+         * Search forward from source
+         */
+        public Map<Boolean, List<Path>> forward(Directions direction) {
```
[the remark this answers is not in the document]
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeTraverser.java]: done

**Pull request closed without merging**